File: mist/src/app/GAMES/project/Enemy.ts
import { Sprite } from './Sprite';
import { Player } from './Player';
import { GamePanel } from './GamePanel';

enum Behavior { TELEPORT = 'teleporter', PATROL = 'patroller', CHASE = 'chaser' }

export class Enemy extends Sprite {
    private behavior: Behavior;
    private player: Player;
    private patrolPath?: [number, number][];
    private idx = 0;
    private speed = 8;
    private teleportCooldown = 100;
    private teleportCounter = 0;

    constructor(
        img: HTMLImageElement,
        x: number,
        y: number,
        player: Player,
        type: string,
        patrolPath?: [number, number][]
    ) {
        super(img, x, y);
        this.player = player;
        this.patrolPath = patrolPath;
        this.behavior = type as Behavior;
    }

    update(panel: GamePanel, collisionEnabled: boolean) {
        if (!collisionEnabled) return;
        if (this.behavior === Behavior.TELEPORT) {
            this.teleportCounter++;
            if (this.teleportCounter >= this.teleportCooldown) {
                this.teleport(panel);
                this.teleportCounter = 0;
            }
            return;
        }
        if (this.behavior === Behavior.CHASE) this.chase();
        else if (this.behavior === Behavior.PATROL) this.patrol();
    }

    private teleport(panel: GamePanel) {
        const L = panel.maze.getMazeLayout();
        let c: number, r: number;
        do {
            c = Math.floor(Math.random() * L[0].length);
            r = Math.floor(Math.random() * L.length);
        } while (L[r][c] !== 0);
        this.x = c * 40;
        this.y = r * 40;
    }

    private chase() {
        if (this.x < this.player.x) this.x += this.speed;
        else if (this.x > this.player.x) this.x -= this.speed;
        if (this.y < this.player.y) this.y += this.speed;
        else if (this.y > this.player.y) this.y -= this.speed;
    }

    private patrol() {
        if (!this.patrolPath) return;
        const [tx, ty] = this.patrolPath[this.idx];
        if (this.x < tx) this.x += this.speed;
        else if (this.x > tx) this.x -= this.speed;
        if (this.y < ty) this.y += this.speed;
        else if (this.y > ty) this.y -= this.speed;
        if (
            Math.abs(this.x - tx) < this.speed &&
            Math.abs(this.y - ty) < this.speed
        ) {
            this.idx = (this.idx + 1) % this.patrolPath.length;
        }
    }
}

File: Enemy.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.mycompany.mazegame;

/**
 *
 * @author lsuti
 */
import java.awt.image.BufferedImage;
import java.util.Random;

public class Enemy extends Sprite {

    private enum B {
        CHASE, PATROL, TELEPORT
    }
    private B behavior;
    private Player player;
    private int[][] path;
    private int idx = 0, speed = 8, tpCooldown = 100, tpCount = 0;

    public Enemy(BufferedImage img, int x, int y,
            Player p, String type, int[][] patrolPath) {
        super(img, x, y);
        this.player = p;
        this.path = patrolPath;
        switch (type) {
            case "chaser":
                behavior = B.CHASE;
                break;
            case "patroller":
                behavior = B.PATROL;
                break;
            default:
                behavior = B.TELEPORT;
                break;
        }
    }

    @Override
    public void update(GamePanel panel, boolean collisionEnabled) {
        if (!collisionEnabled) {
            return;
        }

        if (behavior == B.TELEPORT) {
            if (++tpCount >= tpCooldown) {
                teleport(panel);
                tpCount = 0;
            }
        } else if (behavior == B.CHASE) {
            chase();
        } else {
            patrol();
        }
    }

    private void teleport(GamePanel p) {
        Random r = new Random();
        int[][] L = Maze.getMazeLayout();
        int c, rr;
        do {
            c = r.nextInt(L[0].length);
            rr = r.nextInt(L.length);
        } while (L[rr][c] != 0);
        x = c * 40;
        y = rr * 40;
    }

    private void chase() {
        int px = player.getX(), py = player.getY();
        if (x < px) {
            x += speed;
        } else if (x > px) {
            x -= speed;
        }
        if (y < py) {
            y += speed;
        } else if (y > py) {
            y -= speed;
        }
    }

    private void patrol() {
        if (path == null || path.length == 0) {
            return;
        }
        int tx = path[idx][0], ty = path[idx][1];
        if (x < tx) {
            x += speed;
        } else if (x > tx) {
            x -= speed;
        }
        if (y < ty) {
            y += speed;
        } else if (y > ty) {
            y -= speed;
        }
        if (Math.abs(x - tx) < speed && Math.abs(y - ty) < speed) {
            idx = (idx + 1) % path.length;
        }
    }
}
